src/eventHandlers/checkSuite/index.spec.ts

```diff
@@ -334,8 +336,69 @@ describe('check Suite event handler', (): void => {
         },
       });
 
-    await checkSuiteHandle(octokit, 'dependabot-preview[bot]');
+    await checkSuiteHandle(octokit, 'dependabot-preview[bot]', 3);
 
     expect(warningSpy).toHaveBeenCalled();
   });
+
+  it('retries up to two times before failing', async (): Promise<void> => {
+    expect.assertions(7);
+
+    nock('https://api.github.com')
+      .post('/graphql')
+      .reply(OK, {
+        data: {
+          repository: {
+            pullRequest: {
+              commits: {
+                edges: [
+                  {
+                    node: {
+                      commit: {
+                        messageHeadline: COMMIT_HEADLINE,
+                      },
+                    },
+                  },
+                ],
+              },
+              id: PULL_REQUEST_ID,
+              mergeStateStatus: 'CLEAN',
+              mergeable: 'MERGEABLE',
+              merged: false,
+              reviews: {
+                edges: [
+                  {
+                    node: {
+                      state: 'APPROVED',
+                    },
+                  },
+                ],
+              },
+              state: 'OPEN',
+            },
+          },
+        },
+      })
+      .post('/graphql')
+      .times(3)
+      .reply(403, 'Error when merging');
+
+    const mergeWithRetrySpy = jest.spyOn(merge, 'mergeWithRetry');
+    const logDebugSpy = jest.spyOn(log, 'logDebug');
+    const logInfoSpy = jest.spyOn(log, 'logInfo');
+
+    try {
+      await checkSuiteHandle(octokit, 'dependabot-preview[bot]', 2);
+    } catch (error) {
+      expect(error).toBeInstanceOf(Error);
+      expect(error.message).toStrictEqual('Error when merging');
+      expect(mergeWithRetrySpy).toHaveBeenCalledTimes(3);
```

Comment:
This assertion doesn't seem necessary to me as it is part of the internal implementation.

Reply:
@merlinnot Getting stuck here with @ricardoatsouza. WDYT?

Reply:
Hm, that seems right - we should be able to replace entire implementation without any tests breaking, as long as the behavior presented externally remains the same.

Reply:
@JrSchild pointed out that nock is already ensuring the endpoint to be called `3` times via `times(3)`. That's true, however, for example, if that number is changed to `4`, nock will fail with `Pending mocks detected`, which does not directly reflects the input given for number of retries.

My arguments are:
 - The number of `times` in nock is not related to the number of retries. If the implementation of the `merge` function changes, for example, by calling the endpoint two times instead of one, the number in nock will not anymore coincide with the number of retries.
 - This assertion, though an internal implementation, reflects what is expected of the functionality of the feature: to retry a certain number of times.

Reply:
The test will fail if it doesn't behave correctly, which is what we want.

> The number of times in nock is not related to the number of retries. If the implementation of the merge function changes, for example, by calling the endpoint two times instead of one, the number in nock will not anymore coincide with the number of retries.

The same can be said for the function: it can call itself two times at the same moment if you change the implementation. I don't really see how it changes anything.

What's more, we'll most likely refactor it as soon as our backoff implementation is extracted into a package and open sourced. We currently have an implementation for it used in multiple internal libraries and applications, which is unfortunately copy-pasted.

Reply:
> The test will fail if it doesn't behave correctly, which is what we want.

The test will fail in both cases. My reasoning is towards what is being tested and the reason for a test to fail. What is being tested is the number of calls of a certain function which has to match with the number of retries given as input. The number of cals towards an endpoint is conveniently the same number, but does not mean the same thing.

> What's more, we'll most likely refactor it as soon as our backoff implementation is extracted into a package and open sourced. We currently have an implementation for it used in multiple internal libraries and applications, which is unfortunately copy-pasted.

What does this means in terms of this PR? To proceed and merge once the comments are addressed (and refactor in the future), or to wait for the backoff implementation to be open sourced and then implement it here?

Reply:
> What does this means in terms of this PR? To proceed and merge once the comments are addressed (and refactor in the future), or to wait for the backoff implementation to be open sourced and then implement it here?

That is an argument towards testing external behavior exclusively, as we'll be replacing the implementation. That's always the case, good tests are not dependent on specific implementation, they only test the behavior. Here, we're clearly testing this concrete implementation, which does not allow us to refactor it without refactoring tests. That should never be the case.

Reply:
This line checks whether the already exported function is actually being called certain number of times, which is to me behavior related, and not a specific part of the implementation. Depending on what is being refactored, the output will still be the same.

Reply:
The behavior is already being tested with nock. We have made sure that the endpoint is being called with, and fails exactly `n` times. That allows us to change the implementation any way we want without changing the tests. So as @merlinnot pointed out, there should be no need to test the internal implementation because we have already covered the behavior.

Reply:
> So as @merlinnot pointed out, there should be no need to test the internal implementation because we have already covered the behavior.

I don't agree. The fact that the endpoint is being called exactly `n` times is more of a _coincidence_. The implementation could change in a way that makes the endpoint be called `n + x` times, which would not only make the test fail, but do not reflect the number `n` of retries.

Reply:
Removed in d0b59dc